### ISSUE TITLE
`using` and introspection

### ISSUE THREAD

**Issue by chaoflow** [sandstorm-io/capnproto#411]
```
@0xb575a39b4a533790;

using "/my_framework/types.capnp".Timedelta

struct Foo {
  timedelta @0 Timedelta;
}
```

types.capnp:
```
@0x99ea4b915fea8c65;

using Timedelta = UInt64;
```

Is there any way to find out via introspection, that Foo.timedelta is of type "Timedelta" or are the `using` directives resolved too early for that?

If not, is the filename available from which a schema was loaded from?

**Comment by kentonv** [comment]
At present, `using` defines an alias that is substituted at schema compiling time -- the compiled schemas (in schema.capnp format) do not contain any information about aliases.

I do feel like it makes sense to introduce a second kind of alias, which is intended to be more opaque, and produces type aliases in the generated code. However, I don't have any specific plan in mind... thoughts?

**Comment by chaoflow** [comment]
My use case is type conversion as discussed [here](https://groups.google.com/d/msg/capnproto/YaeLDnShN_c/g4XoL04QS1gJ). An alternative could be annotations for `using` statements as suggested [here](https://groups.google.com/d/msg/capnproto/YaeLDnShN_c/z9cu_v00nCUJ):
```
@0x99ea4b915fea8c65;

annotation type(*) :Text;
using Timedelta = UInt64 $type("timedelta");
```

**Comment by kentonv** [comment]
Yeah, I see what you're going for. Unfortunately currently your only option is to box the timestamp in a struct, which of course wastes 8 bytes on a pointer. :/

**Comment by chaoflow** [comment]
@kentonv Do you have some hints where support for opaque aliases/using directives would need to be implemented. We're happy to customize and evaluate within setup.

**Comment by kentonv** [comment]
@chaoflow This would be fairly involved, requiring some design work, since currently there's no representation of aliases at all in the schema format. I'd want to spec out the design before implementing anything. I expect to have more time to work on Cap'n Proto stuff in a month or two, as I'll be starting a job at a new company that is a heavy user of Cap'n Proto.

I also wonder if your problem might be better-solved by either:
* Adding a built-in `Date` type.
* Implementing the idea of "encapsulated types", in which you can define a wrapper type which uses hand-written types in each language. Perhaps some mechanism of this would allow us to define a `Date` type that becomes a `kj::Date` in C++, a Java `Date` in Java, etc., without needing the type to be built-in...

**Comment by KubaO** [comment]
The main use case for this is the avoidance of boxing of primitive types. A logical generalization of the unboxing would be fixed-size unboxed composite types. To me, a fixed size array begs for a fixed size struct, since they'd be internally handled similarly it seems.

**Comment by kentonv** [comment]
@KubaO We actually had fixed-size inline structs and lists in a very early version of Cap'n Proto, but the feature was removed because it created excessive implementation complexity.

https://groups.google.com/d/msg/capnproto/_jnRqvZeWaY/PoJXz-XQ2PkJ
